internal/bufferedwrites/buffered_write_handler.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufferedwrites
+
+import (
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/googlecloudplatform/gcsfuse/v2/internal/block"
+	"golang.org/x/sync/semaphore"
+)
+
+// Note: All the write operations take inode lock in fs.go, hence we don't need any locks here
+// as we will get write operations serially.
+
+// BufferedWriteHandler is responsible for filling up the buffers with the data
+// as it receives and handing over to uploadHandler which uploads to GCS.
+type BufferedWriteHandler struct {
+	current   block.Block
+	blockPool *block.BlockPool
+	// Total size of data buffered so far. Some part of buffered data might have
+	// been uploaded to GCS as well.
+	totalSize int64
+	// Stores the mtime value updated by kernel as part of setInodeAttributes call.
+	mtime time.Time
+}
+
+// WriteFileInfo is used as part of serving fileInode attributes (GetInodeAttributes call).
+type WriteFileInfo struct {
+	TotalSize int64
+	Mtime     time.Time
+}
+
+// InitBWHandler initialises the bufferedWriteHandler struct.
+func InitBWHandler(blockSize int64, maxBlocks int32, globalMaxBlocksSem *semaphore.Weighted) (bwh *BufferedWriteHandler, err error) {
+	bp, err := block.InitBlockPool(blockSize, maxBlocks, globalMaxBlocksSem)
+	if err != nil {
+		return
+	}
+
+	bwh = &BufferedWriteHandler{
+		current:   nil,
+		blockPool: bp,
+		totalSize: 0,
+		mtime:     time.Now(),
+	}
+	return
+}
+
+// Write writes the given data to the buffer. It writes to an existing buffer if
+// the capacity is available otherwise writes to a new buffer.
+func (wh *BufferedWriteHandler) Write(data []byte, offset int64) (err error) {
+	if offset > wh.totalSize {
+		// TODO: Will be handled as part of ordered writes.
+		return fmt.Errorf("non sequential writes")
+	}
+
+	dataWritten := 0
+	for dataWritten < len(data) {
+		if wh.current == nil {
+			wh.current, err = wh.blockPool.Get()
+			if err != nil {
+				return
```

Comment:
The error should be a bit more descriptive to clarify what went wrong, rather than just forwarding the downstream error. E.g.

```suggestion
				return fmt.Errorf("failed to get current block to write to: %w", err)
```